iac-extensions/kubernetes/src/main/java/org/sonar/iac/helm/tree/utils/GoTemplateAstConverter.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * SonarQube IaC Plugin
+ * Copyright (C) 2021-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.helm.tree.utils;
+
+import com.google.protobuf.Any;
+import com.google.protobuf.AnyOrBuilder;
+import com.google.protobuf.GeneratedMessageV3;
+import com.google.protobuf.Message;
+import com.google.protobuf.MessageOrBuilder;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.annotation.CheckForNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonar.iac.helm.tree.Node;
+
+public final class GoTemplateAstConverter {
+  private static final Logger LOG = LoggerFactory.getLogger(GoTemplateAstConverter.class);
+  private static final MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+  private GoTemplateAstConverter() {
+    // utility class
+  }
+
+  @CheckForNull
+  public static Node unpackNode(Any nodePb) {
+    try {
+      Class<GeneratedMessageV3> messageClass = messageType(nodePb);
+      Class<MessageOrBuilder> messageClassOrBuilder = messageTypeOrBuilder(nodePb);
+      Class<Node> targetClass = treeTypeToMessageType(nodePb);
+      var handle = lookup.findStatic(targetClass, "fromPb", MethodType.methodType(Node.class, messageClassOrBuilder));
```

Comment:
I couldn't think of a good way to make all `Node` subtypes have a common static method. Maybe its presence can be at least checked with archunit

Reply:
I was afraid that you would use some reflection magic. As this code will be called many times during the scanning I would like to avoid using reflection here because of performance reasons.
And the huge switch is still there, so maybe calling the proper method in each `case` is not so bad (the compiler will optimize it well).
Alternatively, you can create a `Map<String, AbstractNode>`, (define factory method as not static, maybe in internal class `Builder` to benefit from inheritance) and call the method without reflection.

Reply:
Changed in 07fd016